cmd/prometheus/main.go

```diff
@@ -94,13 +97,8 @@ func main() {
 
 	a.HelpFlag.Short('h')
 
-	a.Flag("log.level",
-		"Only log messages with the given severity or above. One of: [debug, info, warn, error, fatal]").
-		Default("info").StringVar(&cfg.logLevel)
-
-	a.Flag("log.format",
-		`Set the log target and format. Example: "logger:syslog?appname=bob&local=7" or "logger:stdout?json=true"`).
-		Default("logger:stderr").StringVar(&cfg.logFormat)
+	a.Flag(promlog.LevelFlagName, promlog.LevelFlagHelp).
```

Comment:
> But from your general position I take that you will also object that every component defines a helper type to implement that interface around their flag library.

Actually I'd be okay with that. The goal is that you only have to put in the code to support "Prometheus logging" once, and everything after that is just updating vendoring. Whether that  boilerplate code is 1 line or 10 isn't that important, though shorter is better.

What I'm not okay with is the boilerplate changing.

> This means that components that cannot break anything right now, are stuck in a limbo where they cannot update such a package, even with crucial bug/security fixes, because this update may also pull in a breakage.

This doesn't change with your proposal. It just moves that pain around.

> It's hardly any more effort than updating the vendoring and per my above reasoning, this even should be a manual and conscious step and not a side effect of vendoring.

It is notable effort, as it requires intelligence to deal with. I'm against this requiring any conscious thought in the general case, as that means in practice that noone would ever update. Or maybe they'd choose another library that doesn't impose a maintenance burden every time it gets a new feature, and we end up with inconsistency in the ecosystem.

> The amount of manual(!) effort this requires to not accidentally do this or actually forking/porting partial important fixes temporarily (=~ up to years) is significantly higher than acknowledging that there's a new logging option and adding a flag for it.

Significantly higher effort for whom? We should be considering not just our effort (most of which we're going to have to do anyway), but the knock on effects for developers in the ecosystem using this library and end users. I don't accept the idea that hundreds of developers having to manually review every change to a dependency and update their code for it is less effort than us making a library that doesn't require that review&update.

For me this is just the standard cost of common code, which we will be taking on no matter what we decide. I believe the issues you're concerned about are irreducible complexity, so they'll just show up to bite us and the ecosystem in another form. What I propose makes this cost more manageable.

> The constants we have ensure the important parts of uniformity.

I personally don't like these constants at all, as they add indirection that makes it harder to figure out where and how flags are actually defined.

> I'm saying that because I initiated everything that's the common repository today and in the long run, it has caused nothing but trouble. It was one of my major mistakes in regards to our code base.

I think common has generally worked fine, and is widely used. Some of the code in some of the packages may not be the best, but as common libraries go it meets my expectations.

> It's a very expensive solution that only pays of at a development throughput which we do not remotely have today and will in all likelihood never have.

I'm not seeing this high expense, or why development throughput is related.

From my standpoint one application depending on the internals of another is verboten. The instant that is about to happen, that code should be instead moved out to a common library (which is usually somewhere in prometheus/common for us). Keeping non-trivial (and in many cases trivial) copied&pasted code in sync across 2 applications is not really practical, let alone across tens to hundreds.

Putting common code in the Prometheus repo doesn't stop it from being a common library or change any of the maintenance burdens that having common libraries implies. It just muddles the boundaries between what is considered internal and what is considered external to Prometheus.